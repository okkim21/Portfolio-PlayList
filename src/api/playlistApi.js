const baseUrl = process.env.REACT_APP_API_BASE_URL;

export const getPlaylists = async ()  => {
  const url = baseUrl + "playlists";

  try {
    const response = await fetch(url);
    if (response.ok) {
      const json = await response.json();
      return json;
    } else {
      throw response;
    }
  } catch (e) {
    throw e;
  }
};


export const addPlaylist = async (palyList)  => {
  const url = baseUrl + "playlists";

  try {
    const response = await fetch(url, {
      method : 'POST',
      headers: {
        'Content-type': 'application/json',
      },
      body: JSON.stringify(palyList),
    })

    if (response.ok) {
      const json = await response.json();
      return json;
    } else {
      throw response;
    }
  } catch (e) {
    throw e;
  }
};

export const deletePlayList = async (id)  => {

  const url = baseUrl + `playlists/${id}`;

  try {
    const response = await fetch(url, {
      method : 'DELETE',
    });

    if (response.status !== 200) {
      throw response;
    }
  } catch (e) {
    throw e;
  }
};