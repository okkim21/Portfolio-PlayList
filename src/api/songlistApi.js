const baseUrl = process.env.REACT_APP_API_BASE_URL;


export const getSonglists = async (id)  => {

  const url = baseUrl + `playlists/${id}`;

  try {
    const response = await fetch(url);
    if (response.ok) {
      const json = await response.json();
      return json;
    } else {
      throw response;
    }
  } catch (e) {
    throw e;
  }
};

