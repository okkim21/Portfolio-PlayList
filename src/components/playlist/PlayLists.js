import { useState, useEffect } from "react";
import { getPlaylists, addPlaylist, deletePlayList } from "../../api/playlistApi";
import { Link, useRouteMatch } from "react-router-dom";
import { FaTimes } from 'react-icons/fa'
import SongTitles from "./SongTitles";
import AddPlayList from "./AddPlayList";

function PlayLists() {
  const [playLists, setPlayLists] = useState([]);
  const [showAddPlayList, setShowAddPlayList] = useState(false);
  let { url } = useRouteMatch();

  useEffect(() => {
    async function fetchPlayLists() {
      const playLists = await getPlaylists();
      setPlayLists(playLists);
    }

    fetchPlayLists();
  }, []);

  function renderPlayLists(p) {
 
    return (
      <div id="playList_container" key={p.id}>
        <header>
          <h2 id="name">{p.name}</h2>
          <p>{p.date}</p>
          <FaTimes
          onClick={()=>deletePlay(p.id)}
          style={{ color: 'red', cursor: 'pointer' }}/>
        </header>
        <Link to={`${url}/${p.id}`}>
          <img src={`/images/${p.image}`} alt={p.name} />
        </Link>
        <section>
        { p.songLists.length > 0 && <SongTitles songLists={p.songLists} />}
        </section>
      
      </div>
    );
  }

  // Add PalyList
  const addPlay = async (playlist) => {
    const data = await addPlaylist(playlist);
    setPlayLists([...playLists, data])
    setShowAddPlayList(false)
  }

  // Delete PalyList
  const deletePlay = async (id) => {
    await deletePlayList(id);
    setPlayLists(playLists.filter((p) => p.id !== id))
  }

  return (
    <>
      <button className="btn" onClick={()=> setShowAddPlayList(!showAddPlayList)} >
      {showAddPlayList ? "Close" : "Add PlayList"}
      </button>
      {showAddPlayList && <AddPlayList onAdd={addPlay} />}
      <section className="playLists">{playLists.map(renderPlayLists)}</section>
    </>
  );
}

export default PlayLists;
