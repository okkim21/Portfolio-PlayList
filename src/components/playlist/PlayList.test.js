import * as React from "react";
import PlayLists from "./PlayLists";
import { act } from "react-dom/test-utils";
import * as playlistApi from "../../api/playlistApi";
import { shallow } from "enzyme";


jest.mock('react-router-dom', () => ({
  __esModule: true,
  useRouteMatch: () => ({ url: '/playlists' }),
}));

describe("rendering PalyList component", () => {
  let data = {
    id: 1,
    name: "Spring_PlayList",
    date: "03/01/2020",
    image: "album.jpg",
    songLists: [],
  };
 
  it("renders PlayList Component without crashing", () => {
    const wrapper = shallow(<PlayLists />);
    expect(wrapper).toMatchSnapshot();
  })

  it("should show palylist once async resolves", async () => {
    await act(async () => {
        const spy_getPlaylists = jest.spyOn(playlistApi, "getPlaylists").mockImplementation(()=> Promise.resolve({
          status: 200,
          data,
          json: () => data
        }))
        const result = await spy_getPlaylists();
        expect(result.data.id).toBe(1);
        spy_getPlaylists.mockRestore();
    
    });
  });

});
