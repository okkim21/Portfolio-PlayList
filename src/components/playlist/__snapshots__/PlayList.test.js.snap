// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering PalyList component renders PlayList Component without crashing 1`] = `ShallowWrapper {}`;

exports[`rendering component renders PlayList Component without crashing 1`] = `ShallowWrapper {}`;
